bundles/org.openhab.persistence.jpa/src/main/java/org/openhab/persistence/jpa/internal/JpaHistoricItem.java

```diff
@@ -105,8 +107,21 @@ public static HistoricItem fromPersistedItem(JpaPersistentItem pItem, Item item)
         State state;
         if (item instanceof NumberItem numberItem) {
             Unit<?> unit = numberItem.getUnit();
-            double value = Double.parseDouble(pItem.getValue());
-            state = (unit == null) ? new DecimalType(value) : new QuantityType<>(value, unit);
+            QuantityType<?> value = QuantityType.valueOf(pItem.getValue());
+            if (unit == null) {
+                // Item has no unit; drop any persisted unit
+                state = Objects.requireNonNull(value.as(DecimalType.class));
+            } else if (value.getUnit() == Units.ONE) {
+                // No persisted unit; assume the item's unit
```

Comment:
This used to be the case, but I believe the exception for Unit.ONE may have been removed. So depending on when it was stored, it could have ‘one’ behind it.

Reply:
That doesn't really matter though. If it has "one" or not, it will hit this case. If the item is dimensioned, it will return the value with the item's unit (which _might_ be "one"!). If it's not, we'll have already hit the case above and returned a DecimalType.

Reply:
It will matter if for one reason or the other it is stored with another dimensionless unit than Unit.ONE and the current item unit is Unit.ONE. Chances are small for this, but not impossible.

Reply:
Let's walk through this.
 * I start with an item, `Number:Dimensionless MyItem { unit="%" }`.
 * It gets a state `50 %` (as a `QuantityType`), which is then persisted to the database as the string `"50 %"`.
 * The item then changes definition to `Number:Dimensionless MyItem` (no unit, thus assumes the default of `one` for the Dimensionless dimension.)
 * A rule or script or something calls `persistedState` for `Myitem`.
 * `50 %` gets parsed by `QuantityType` as 50, with unit `%`.
 * The unit on the item is `one`, not `null`, so that branch isn't taken.
 * The unit on the parsed value is `%`, not `one`, so this branch isn't taken.
 * The else branch is taken. `%` _is_ dimensionless, and compatible with `one`, so gets converted, and returned as `0.5` (with unit `one`).

If we start with other dimensionless units, the results are as follows:
 * `50 ppm` => `0.00005` (one)
 * `50 dB` => `100000.0000000002` (one)
I'm not sure if any of those are really problematic.

Let's try the other way around. Start as `one`, and item changes to `%`.
 * The state gets persisted as `50`
 * `50` gets parsed as 50, with unit `one`
 * This branch is taken in the `if`
 * It will construct a quantity type `50 %` (or `50 ppm` or `50 dB` for other dimensionless units; or `50 gal` for even "incompatible" dimensioned units)

Note that this case is the exact same if the item started out as a `Number` (without dimension/unit). If it started out as a `Dimmer`, a `PercentType` would be serialized as `50.0`, and thus the current code would correctly reproduce the original semantic state.

I just don't see any way to distinguish between "serialized as `one`, so need to apply the inverse conversion from `one` to the current unit" vs. "serialized as a number without any dimension, and it should just blindly apply the current unit". Both cases serialize a number (as a string) without any additional information.

> This used to be the case, but I believe the exception for Unit.ONE may have been removed.

I assume you're talking about https://github.com/openhab/openhab-core/pull/4169? I don't see that the change to toFullString() in that PR makes any difference - `one` is not actually included in the output either way. Maybe quantity serialization was different in the past? In particular, I'm looking at `QuantityType.toString`, which is how JPA serializes `QuantityType`s, and not `QuantityType.format`, which may include(d) a state description which may have forcibly included the unit symbol.

Reply:
QuantityType.toString calls QuantityType.toFullString, and that has been changed in the PR you reference, removing the exception where only the value was output for Unit.ONE.

Are there cases where old persisted data would not have a unit (probably from before the introduction of QuantityType)? If so, interpreting no stored unit as the unit of the item could make sense. In all other cases, I would advocate for not returning something if the stored unit is not compatible with the item unit. I don’t think you need a special treatment for Unit.ONE then.

Reply:
> QuantityType.toString calls QuantityType.toFullString, and that has been changed in the PR you reference, removing the exception where only the value was output for Unit.ONE.

I don't think that change actually did anything, though. In current 4.3 snapshot (so, with that PR), `QuantityType.toFullString` _still_ does not include `one` in its result, even if the unit is one.

> Are there cases where old persisted data would not have a unit (probably from before the introduction of QuantityType)?
Anytime a user has persistence configured, and changes an item from bare `Number` to `Number:<some dimension>`. Which I think is a pretty common use case when a user realizes that UoM is a thing, and will help ease working with a binding that natively works in units different than their preferences, so they go around tagging as many items as they can find with the proper units.

Reply:
> Anytime a user has persistence configured, and changes an item from bare `Number` to `Number:<some dimension>`. Which I think is a pretty common use case when a user realizes that UoM is a thing, and will help ease working with a binding that natively works in units different than their preferences, so they go around tagging as many items as they can find with the proper units.

That's a very good point indeed, and justifies the extra handling for Units.ONE.

> I don't think that change actually did anything, though. In current 4.3 snapshot (so, with that PR), `QuantityType.toFullString` _still_ does not include `one` in its result, even if the unit is one.

Interesting observation. I am at a loss now, because it did solve https://github.com/openhab/openhab-core/issues/4166. Looking at the indriya code, I can only conclude that AbstractUnit.ONE is created with an empty symbol string (https://github.com/unitsofmeasurement/indriya/blob/7a334700634ee1318711098bf9745e6c348dbb69/src/main/java/tech/units/indriya/AbstractUnit.java#L107).

Reply:
Is there anything more to add to this conversation? If not, i'll merge this as is.

Reply:
Nothing to add.